compiler/src/main/java/com/squareup/anvil/compiler/codegen/BindingModuleGenerator.kt

```diff
@@ -190,67 +188,135 @@ internal class BindingModuleGenerator(
+          .filter {
+            val annotation = it.annotationOrNull(annotationFqName)
+            annotation != null && scope == annotation.scope(module).fqNameSafe
+          }
+          .map { contributedClass ->
+            val annotation = contributedClass.annotation(annotationFqName)
+            val boundType = annotation.boundType(module, contributedClass, isMultibinding)
+
+            checkExtendsBoundType(type = contributedClass, boundType = boundType)
+            checkNotGeneric(type = contributedClass, boundTypeDescriptor = boundType)
+
+            val concreteType = contributedClass.fqNameSafe
+
+            val qualifiers = if (annotation.ignoreQualifier()) {
+              emptyList()
+            } else {
+              contributedClass.annotations
+                .filter { it.isQualifier() }
+                .map { it.toAnnotationSpec(module) }
+            }
+
+            if (DescriptorUtils.isObject(contributedClass)) {
+              ProviderMethod(
+                FunSpec
+                  .builder(
+                    name = concreteType
+                      .asString()
+                      .split(".")
+                      .joinToString(
+                        separator = "",
+                        prefix = "provide",
+                        postfix = if (isMultibinding) "Multi" else ""
+                      ) {
+                        it.capitalize(US)
+                      }
+                  )
+                  .addAnnotation(Provides::class)
+                  .apply {
+                    if (isMultibinding) {
+                      addAnnotation(IntoSet::class)
                     }
-                )
-                .addAnnotation(Provides::class)
-                .addAnnotations(qualifiers)
-                .returns(boundType.asClassName())
-                .addStatement("return %T", contributedClass.asClassName())
-                .build()
-            )
-          } else {
-            BindingMethod(
-              FunSpec
-                .builder(
-                  name = concreteType
-                    .asString()
-                    .split(".")
-                    .joinToString(separator = "", prefix = "bind") { it.capitalize(US) }
-                )
-                .addAnnotation(Binds::class)
-                .addAnnotations(qualifiers)
-                .addModifiers(ABSTRACT)
-                .addParameter(
-                  name = concreteType.shortName().asString().decapitalize(US),
-                  type = contributedClass.asClassName()
-                )
-                .returns(boundType.asClassName())
-                .build()
-            )
+                  }
+                  .addAnnotations(qualifiers)
+                  .returns(boundType.asClassName())
+                  .addStatement("return %T", contributedClass.asClassName())
+                  .build()
+              )
+            } else {
+              BindingMethod(
+                FunSpec
```

Comment:
This API is quite nice. You can visualize how the function looks in code easily by following this chain.

Reply:
I love Kotlin Poet, it's one of the few libraries that still feels magical to me.